### PR TITLE
disttask: record endtime for task and subtask

### LINKED ISSUE #49617: record subtask end time
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 3. What did you see instead (Required)

image

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: close #49617 

Problem Summary:

### What changed and how does it work?
1. when task is done, record end time for it.
2. move task state related table op into one file.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/disttask/framework/storage/table_test.go [hunk 1]
```diff
@@ -131,11 +131,15 @@ func TestTaskTable(t *testing.T) {
...
 	require.Nil(t, task.Error)
+	curTime := time.Unix(time.Now().Unix(), 0)
 	require.NoError(t, gm.FailTask(ctx, id, proto.TaskStatePending, errors.New("test error")))
 	task, err = gm.GetTaskByID(ctx, id)
 	require.NoError(t, err)
 	require.Equal(t, proto.TaskStateFailed, task.State)
 	require.ErrorContains(t, task.Error, "test error")
+	endTime, err := storage.GetTaskEndTimeForTest(ctx, gm, id)
+	require.NoError(t, err)
+	require.GreaterOrEqual(t, endTime, curTime)
```
Comment (okJiang) [inline pkg/disttask/framework/storage/table_test.go:142]: Checking if their difference is smaller than a deviation?

**Review by okJiang — COMMENTED** [review]
rest lgtm

File: pkg/disttask/framework/storage/table_test.go (hunk 1 above)
Comment (ywqzzy) [reply pkg/disttask/framework/storage/table_test.go:142]:
```
	require.LessOrEqual(t, endTime.After(curTime), time.Now().After(curTime))

```
add this?

Comment (ywqzzy) [reply pkg/disttask/framework/storage/table_test.go:142]: Add this?
```
	require.LessOrEqual(t, endTime.Sub(curTime), time.Now().Sub(curTime))

```

Comment (okJiang) [reply pkg/disttask/framework/storage/table_test.go:143]: I mean endTime-curTime < 1s

File: pkg/disttask/framework/storage/task_table.go [hunk 2]
```diff
@@ -413,7 +413,8 @@ func (stm *TaskManager) FailTask(ctx context.Context, taskID int64, currentState
 		`update mysql.tidb_global_task
 			set state=%?,
 			    error = %?,
-			    state_update_time = CURRENT_TIMESTAMP()
+			    state_update_time = CURRENT_TIMESTAMP(),
+				end_time = CURRENT_TIMESTAMP()
```
Comment (D3Hunter) [inline pkg/disttask/framework/storage/task_table.go:417]: some terminate state of task is updated using `UpdateTaskAndAddSubTasks`, see where `.OnDone(` is called

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: pkg/disttask/framework/scheduler/scheduler.go [hunk 3]
```diff
@@ -641,12 +641,12 @@ func (s *BaseScheduler) handlePlanErr(err error) error {
 	if s.IsRetryableErr(err) {
 		return err
 	}
-	s.Task.Error = err
```
Comment (D3Hunter) [inline pkg/disttask/framework/scheduler/scheduler.go:644]: this is required, `OnDone` use this to check whether it's success or failed

**Review by D3Hunter — APPROVED** [review]

File: pkg/disttask/framework/storage/table_test.go (hunk 1 above)
Comment (ywqzzy) [reply pkg/disttask/framework/storage/table_test.go:145]: We can't make sure endTime-curTime < 1s

File: pkg/disttask/framework/scheduler/interface.go [hunk 4]
```diff
@@ -36,10 +36,18 @@ type TaskManager interface {
 	GetAllNodes(ctx context.Context) ([]proto.ManagedNode, error)
 	DeleteDeadNodes(ctx context.Context, nodes []string) error
 	TransferTasks2History(ctx context.Context, tasks []*proto.Task) error
+	// CancelTask updated task state to canceling.
 	CancelTask(ctx context.Context, taskID int64) error
 	// FailTask updates task state to Failed and updates task error.
 	FailTask(ctx context.Context, taskID int64, currentState proto.TaskState, taskErr error) error
+	// RevertedTask updates task state to reverted.
+	RevertedTask(ctx context.Context, taskID int64) error
+	// PausingTask updated task state to pausing.
```
Comment (ywqzzy) [inline pkg/disttask/framework/scheduler/interface.go:45]:
```suggestion
	// PauseTask updated task state to pausing.
```

**Review by okJiang — APPROVED** [review]

**Pull request merged**